File: services/ads-api/internal/service/rabbitmq.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"time"

	"github.com/rabbitmq/amqp091-go"
)

type RabbitMQService struct {
	conn    *amqp091.Connection
	channel *amqp091.Channel
}

func NewRabbitMQService(url string) (*RabbitMQService, error) {
	conn, err := amqp091.Dial(url)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to RabbitMQ: %w", err)
	}

	ch, err := conn.Channel()
	if err != nil {
		return nil, fmt.Errorf("failed to open channel: %w", err)
	}

	err = ch.ExchangeDeclare(
		"ad_events",
		"topic",
		true,
		false,
		false,
		false,
		nil,
	)
	if err != nil {
		return nil, fmt.Errorf("failed to declare exchange: %w", err)
	}

	return &RabbitMQService{
		conn:    conn,
		channel: ch,
	}, nil
}

func (r *RabbitMQService) PublishAdCreated(ctx context.Context, adID string) error {
	event := map[string]interface{}{
		"type":    "ad_created",
		"ad_id":   adID,
		"created": time.Now().Format(time.RFC3339),
	}

	body, err := json.Marshal(event)
	if err != nil {
		return fmt.Errorf("failed to marshal event: %w", err)
	}

	err = r.channel.PublishWithContext(ctx,
		"ad_events",
		"ad.created",
		false,
		false,
		amqp091.Publishing{
			ContentType: "application/json",
			Body:        body,
		})
	if err != nil {
		return fmt.Errorf("failed to publish message: %w", err)
	}

	log.Printf("Published ad_created event for ad %s", adID)
	return nil
}

func (r *RabbitMQService) PublishAdUpdated(ctx context.Context, adID string) error {
	event := map[string]interface{}{
		"type":    "ad_updated",
		"ad_id":   adID,
		"updated": time.Now().Format(time.RFC3339),
	}

	body, err := json.Marshal(event)
	if err != nil {
		return fmt.Errorf("failed to marshal event: %w", err)
	}

	err = r.channel.PublishWithContext(ctx,
		"ad_events",
		"ad.updated",
		false,
		false,
		amqp091.Publishing{
			ContentType: "application/json",
			Body:        body,
		})
	if err != nil {
		return fmt.Errorf("failed to publish message: %w", err)
	}

	log.Printf("Published ad_updated event for ad %s", adID)
	return nil
}

func (r *RabbitMQService) Close() error {
	if err := r.channel.Close(); err != nil {
		return fmt.Errorf("failed to close channel: %w", err)
	}
	if err := r.conn.Close(); err != nil {
		return fmt.Errorf("failed to close connection: %w", err)
	}
	return nil
}

File: services/auth-api/app/core/rabbitmq.py
import pika
import json
import logging
from typing import Optional
from contextlib import contextmanager

logger = logging.getLogger(__name__)


class RabbitMQService:
    def __init__(self, url: str = "amqp://guest:guest@rabbitmq:5672/"):
        self.url = url
        self.connection = None
        self.channel = None

    def connect(self):
        """Establish connection to RabbitMQ"""
        try:
            self.connection = pika.BlockingConnection(
                pika.URLParameters(self.url)
            )
            self.channel = self.connection.channel()
            
            self.channel.exchange_declare(
                exchange='ad_events',
                exchange_type='topic',
                durable=True
            )
            
            logger.info("Connected to RabbitMQ successfully")
            return True
        except Exception as e:
            logger.error(f"Failed to connect to RabbitMQ: {e}")
            return False

    def publish_ad_created(self, ad_id: str):
        """Publish ad created event"""
        if not self.channel:
            logger.warning("RabbitMQ channel not available")
            return

        event = {
            "type": "ad_created",
            "ad_id": ad_id,
            "created": "2026-03-06T13:00:00Z"
        }

        try:
            self.channel.basic_publish(
                exchange='ad_events',
                routing_key='ad.created',
                body=json.dumps(event),
                properties=pika.BasicProperties(
                    content_type='application/json',
                    delivery_mode=2
                )
            )
            logger.info(f"Published ad_created event for ad {ad_id}")
        except Exception as e:
            logger.error(f"Failed to publish ad_created event: {e}")

    def publish_ad_updated(self, ad_id: str):
        """Publish ad updated event"""
        if not self.channel:
            logger.warning("RabbitMQ channel not available")
            return

        event = {
            "type": "ad_updated",
            "ad_id": ad_id,
            "updated": "2026-03-06T13:00:00Z"
        }

        try:
            self.channel.basic_publish(
                exchange='ad_events',
                routing_key='ad.updated',
                body=json.dumps(event),
                properties=pika.BasicProperties(
                    content_type='application/json',
                    delivery_mode=2
                )
            )
            logger.info(f"Published ad_updated event for ad {ad_id}")
        except Exception as e:
            logger.error(f"Failed to publish ad_updated event: {e}")

    def close(self):
        """Close RabbitMQ connection"""
        if self.channel:
            self.channel.close()
        if self.connection:
            self.connection.close()
        logger.info("RabbitMQ connection closed")


rabbitmq_service = RabbitMQService()


@contextmanager
def get_rabbitmq():
    """Context manager for RabbitMQ connection"""
    if rabbitmq_service.connect():
        try:
            yield rabbitmq_service
        finally:
            rabbitmq_service.close()
    else:
        yield None